File: examples/python/basic/copilot_examples.py
#!/usr/bin/env python3
"""
Ejemplos básicos de Python con GitHub Copilot
Demuestra cómo escribir prompts efectivos para generar código Python
"""

import asyncio
import json
import math
import re
import requests
from datetime import datetime, timedelta
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass
from pathlib import Path


@dataclass
class Person:
    """
    Clase Person generada con GitHub Copilot
    Prompt: Clase que representa una persona con nombre, edad y email
    Debe validar el email y calcular el año de nacimiento
    """
    name: str
    age: int
    email: str
    
    def __post_init__(self):
        """Valida los datos después de la inicialización"""
        if not self.is_valid_email(self.email):
            raise ValueError(f"Email inválido: {self.email}")
        if self.age < 0 or self.age > 150:
            raise ValueError(f"Edad inválida: {self.age}")
    
    @staticmethod
    def is_valid_email(email: str) -> bool:
        """Valida formato de email usando regex"""
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None
    
    @property
    def birth_year(self) -> int:
        """Calcula el año de nacimiento aproximado"""
        return datetime.now().year - self.age
    
    def to_dict(self) -> Dict[str, any]:
        """Convierte la instancia a diccionario"""
        return {
            'name': self.name,
            'age': self.age,
            'email': self.email,
            'birth_year': self.birth_year
        }


class DataProcessor:
    """
    Clase para procesamiento de datos generada con Copilot
    Prompt: Clase que procesa listas de números y texto
    Debe incluir métodos para estadísticas, filtrado y transformación
    """
    
    def __init__(self, data: List[any] = None):
        self.data = data or []
    
    def add_data(self, item: any) -> None:
        """Añade un elemento a los datos"""
        self.data.append(item)
    
    def filter_numbers(self) -> List[float]:
        """Filtra solo los números de los datos"""
        return [item for item in self.data if isinstance(item, (int, float))]
    
    def filter_strings(self) -> List[str]:
        """Filtra solo las strings de los datos"""
        return [item for item in self.data if isinstance(item, str)]
    
    def get_statistics(self) -> Dict[str, float]:
        """Calcula estadísticas básicas de los números"""
        numbers = self.filter_numbers()
        if not numbers:
            return {'count': 0, 'sum': 0, 'mean': 0, 'min': 0, 'max': 0, 'std': 0}
        
        count = len(numbers)
        total = sum(numbers)
        mean = total / count
        variance = sum((x - mean) ** 2 for x in numbers) / count
        std = math.sqrt(variance)
        
        return {
            'count': count,
            'sum': total,
            'mean': mean,
            'min': min(numbers),
            'max': max(numbers),
            'std': std
        }
    
    def word_frequency(self) -> Dict[str, int]:
        """Calcula la frecuencia de palabras en las strings"""
        strings = self.filter_strings()
        word_count = {}
        
        for text in strings:
            words = re.findall(r'\b\w+\b', text.lower())
            for word in words:
                word_count[word] = word_count.get(word, 0) + 1
        
        return dict(sorted(word_count.items(), key=lambda x: x[1], reverse=True))


def fibonacci_sequence(n: int) -> List[int]:
    """
    Prompt: Función que genera la secuencia de Fibonacci hasta n términos
    Debe manejar casos edge como n <= 0 y ser eficiente
    """
    if n <= 0:
        return []
    elif n == 1:
        return [0]
    elif n == 2:
        return [0, 1]
    
    sequence = [0, 1]
    for i in range(2, n):
        next_num = sequence[i-1] + sequence[i-2]
        sequence.append(next_num)
    
    return sequence


def is_palindrome(text: str) -> bool:
    """
    Prompt: Función que determina si un texto es palíndromo
    Debe ignorar espacios, puntuación y mayúsculas
    """
    # Limpiar el texto: solo letras y números, convertir a minúsculas
    cleaned = re.sub(r'[^a-zA-Z0-9]', '', text.lower())
    return cleaned == cleaned[::-1]


def prime_factors(n: int) -> List[int]:
    """
    Prompt: Función que encuentra los factores primos de un número
    Debe ser eficiente y manejar números grandes
    """
    if n <= 1:
        return []
    
    factors = []
    d = 2
    
    while d * d <= n:
        while n % d == 0:
            factors.append(d)
            n //= d
        d += 1
    
    if n > 1:
        factors.append(n)
    
    return factors


async def fetch_url_async(url: str, timeout: int = 10) -> Dict[str, any]:
    """
    Prompt: Función async que hace petición HTTP y retorna información
    Debe manejar timeouts, errores HTTP y parsear JSON si es posible
    """
    try:
        response = requests.get(url, timeout=timeout)
        response.raise_for_status()
        
        result = {
            'url': url,
            'status_code': response.status_code,
            'headers': dict(response.headers),
            'content_length': len(response.content),
            'timestamp': datetime.now().isoformat()
        }
        
        # Intentar parsear JSON
        try:
            result['json_data'] = response.json()
        except json.JSONDecodeError:
            result['text_preview'] = response.text[:200] + '...' if len(response.text) > 200 else response.text
        
        return result
        
    except requests.exceptions.Timeout:
        return {'error': 'Timeout', 'url': url}
    except requests.exceptions.RequestException as e:
        return {'error': str(e), 'url': url}


def analyze_text(text: str) -> Dict[str, any]:
    """
    Prompt: Función que analiza un texto y retorna estadísticas
    Debe contar caracteres, palabras, oraciones y frecuencia de letras
    """
    if not text:
        return {'error': 'Texto vacío'}
    
    # Estadísticas básicas
    char_count = len(text)
    word_count = len(re.findall(r'\b\w+\b', text))
    sentence_count = len(re.findall(r'[.!?]+', text))
    paragraph_count = len([p for p in text.split('\n\n') if p.strip()])
    
    # Frecuencia de letras
    letter_freq = {}
    for char in text.lower():
        if char.isalpha():
            letter_freq[char] = letter_freq.get(char, 0) + 1
    
    # Palabras más frecuentes
    words = re.findall(r'\b\w+\b', text.lower())
    word_freq = {}
    for word in words:
        word_freq[word] = word_freq.get(word, 0) + 1
    
    top_words = sorted(word_freq.items(), key=lambda x: x[1], reverse=True)[:10]
    
    return {
        'characters': char_count,
        'words': word_count,
        'sentences': sentence_count,
        'paragraphs': paragraph_count,
        'average_words_per_sentence': word_count / sentence_count if sentence_count > 0 else 0,
        'letter_frequency': dict(sorted(letter_freq.items(), key=lambda x: x[1], reverse=True)),
        'top_words': top_words
    }


def file_operations_example(file_path: str, data: Dict[str, any]) -> bool:
    """
    Prompt: Función que demuestra operaciones con archivos
    Debe escribir JSON, leer y validar el contenido
    """
    try:
        path = Path(file_path)
        
        # Crear directorio si no existe
        path.parent.mkdir(parents=True, exist_ok=True)
        
        # Escribir datos a archivo JSON
        with open(path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        
        # Leer y verificar
        with open(path, 'r', encoding='utf-8') as f:
            loaded_data = json.load(f)
        
        # Validar que los datos son iguales
        return loaded_data == data
        
    except (IOError, json.JSONDecodeError) as e:
        print(f"Error en operaciones de archivo: {e}")
        return False


async def main():
    """Función principal que demuestra todos los ejemplos"""
    print("🐍 Ejemplos básicos de Python con GitHub Copilot")
    print("=" * 60)
    
    # Ejemplo 1: Clase Person
    print("\n👤 Ejemplo 1: Clase Person")
    try:
        person = Person("Ana García", 25, "ana@example.com")
        print(f"Persona: {person.name}, Edad: {person.age}, Año nacimiento: {person.birth_year}")
        print(f"Dict: {person.to_dict()}")
    except ValueError as e:
        print(f"Error: {e}")
    
    # Ejemplo 2: Procesamiento de datos
    print("\n📊 Ejemplo 2: Procesamiento de datos")
    processor = DataProcessor([1, 2, 3, 4, 5, "hello", "world", "hello", 6.5, "python"])
    stats = processor.get_statistics()
    word_freq = processor.word_frequency()
    print(f"Estadísticas numéricas: {stats}")
    print(f"Frecuencia de palabras: {word_freq}")
    
    # Ejemplo 3: Fibonacci
    print("\n🔢 Ejemplo 3: Secuencia Fibonacci")
    fib = fibonacci_sequence(10)
    print(f"Fibonacci (10 términos): {fib}")
    
    # Ejemplo 4: Palíndromo
    print("\n🔄 Ejemplo 4: Detector de palíndromos")
    test_strings = ["A man a plan a canal Panama", "hello", "racecar", "12321"]
    for text in test_strings:
        result = is_palindrome(text)
        print(f"'{text}' es palíndromo: {result}")
    
    # Ejemplo 5: Factores primos
    print("\n✂️ Ejemplo 5: Factores primos")
    numbers = [12, 17, 100, 97]
    for num in numbers:
        factors = prime_factors(num)
        print(f"Factores primos de {num}: {factors}")
    
    # Ejemplo 6: Petición HTTP async
    print("\n🌐 Ejemplo 6: Petición HTTP asíncrona")
    url_result = await fetch_url_async("https://httpbin.org/json")
    print(f"Resultado de petición: {url_result.get('status_code', 'Error')}")
    
    # Ejemplo 7: Análisis de texto
    print("\n📝 Ejemplo 7: Análisis de texto")
    sample_text = """
    GitHub Copilot es una herramienta revolucionaria para desarrolladores.
    Utiliza inteligencia artificial para ayudar en la programación.
    Es especialmente útil para generar código y acelerar el desarrollo.
    """
    analysis = analyze_text(sample_text)
    print(f"Análisis: {analysis['words']} palabras, {analysis['sentences']} oraciones")
    print(f"Top palabras: {analysis['top_words'][:5]}")
    
    # Ejemplo 8: Operaciones con archivos
    print("\n💾 Ejemplo 8: Operaciones con archivos")
    test_data = {
        "timestamp": datetime.now().isoformat(),
        "examples_completed": 8,
        "framework": "GitHub Copilot + Python"
    }
    
    success = file_operations_example("/tmp/copilot_test.json", test_data)
    print(f"Operaciones de archivo completadas: {success}")
    
    print("\n✅ Todos los ejemplos completados!")


if __name__ == "__main__":
    # Ejecutar ejemplos
    asyncio.run(main())

File: examples/java/basic/CopilotExamples.java
package com.example.copilot;

import java.io.*;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Ejemplos básicos de Java con GitHub Copilot
 * Demuestra cómo escribir prompts efectivos para generar código Java
 */
public class CopilotExamples {
    
    public static void main(String[] args) {
        System.out.println("☕ Ejemplos básicos de Java con GitHub Copilot");
        System.out.println("=".repeat(60));
        
        try {
            // Ejemplo 1: Procesamiento de datos con Streams
            demonstrateStreamProcessing();
            
            // Ejemplo 2: Clases y POO
            demonstrateObjectOrientation();
            
            // Ejemplo 3: Manejo de archivos
            demonstrateFileOperations();
            
            // Ejemplo 4: HTTP Client
            demonstrateHttpClient();
            
            // Ejemplo 5: Programación funcional
            demonstrateFunctionalProgramming();
            
            System.out.println("\n✅ Todos los ejemplos completados!");
            
        } catch (Exception e) {
            System.err.println("❌ Error ejecutando ejemplos: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Prompt: Método que demuestra procesamiento de datos con Streams API
     * Debe incluir filtrado, mapping, agrupación y estadísticas
     */
    private static void demonstrateStreamProcessing() {
        System.out.println("\n📊 Ejemplo 1: Procesamiento de datos con Streams");
        
        // Crear datos de ejemplo
        List<Person> people = Arrays.asList(
            new Person("Ana", 25, "Engineering", 70000),
            new Person("Carlos", 30, "Marketing", 55000),
            new Person("María", 28, "Engineering", 72000),
            new Person("Juan", 35, "Sales", 48000),
            new Person("Laura", 27, "Engineering", 68000),
            new Person("Pedro", 32, "Marketing", 58000)
        );
        
        // Estadísticas por departamento
        Map<String, DoubleSummaryStatistics> salaryStatsByDept = people.stream()
            .collect(Collectors.groupingBy(
                Person::getDepartment,
                Collectors.summarizingDouble(Person::getSalary)
            ));
        
        System.out.println("Estadísticas salariales por departamento:");
        salaryStatsByDept.forEach((dept, stats) -> {
            System.out.printf("  %s: Promedio=%.2f, Min=%.2f, Max=%.2f, Count=%d%n",
                dept, stats.getAverage(), stats.getMin(), stats.getMax(), stats.getCount());
        });
        
        // Top 3 salarios más altos
        List<Person> topEarners = people.stream()
            .sorted(Comparator.comparingDouble(Person::getSalary).reversed())
            .limit(3)
            .collect(Collectors.toList());
        
        System.out.println("\nTop 3 salarios más altos:");
        topEarners.forEach(p -> System.out.printf("  %s: $%.2f%n", p.getName(), p.getSalary()));
        
        // Generar números primos usando Streams
        List<Integer> primes = IntStream.rangeClosed(2, 100)
            .filter(CopilotExamples::isPrime)
            .boxed()
            .collect(Collectors.toList());
        
        System.out.println("\nPrimeros 10 números primos: " + primes.subList(0, 10));
    }
    
    /**
     * Prompt: Método que verifica si un número es primo
     * Debe ser eficiente para números grandes
     */
    private static boolean isPrime(int n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        
        for (int i = 5; i * i <= n; i += 6) {
            if (n % i == 0 || n % (i + 2) == 0) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Ejemplo de programación orientada a objetos
     */
    private static void demonstrateObjectOrientation() {
        System.out.println("\n🏗️ Ejemplo 2: Programación orientada a objetos");
        
        // Crear biblioteca con libros
        Library library = new Library("Biblioteca Central");
        
        library.addBook(new Book("1984", "George Orwell", "978-0451524935", 1949));
        library.addBook(new Book("Clean Code", "Robert Martin", "978-0132350884", 2008));
        library.addBook(new Book("Design Patterns", "Gang of Four", "978-0201633612", 1994));
        
        System.out.println("Biblioteca creada: " + library.getName());
        System.out.println("Total de libros: " + library.getTotalBooks());
        
        // Buscar libros
        List<Book> techBooks = library.searchByKeyword("code");
        System.out.println("\nLibros relacionados con 'code':");
        techBooks.forEach(book -> System.out.println("  - " + book.getTitle()));
        
        // Libros por década
        Map<Integer, List<Book>> booksByDecade = library.getBooks().stream()
            .collect(Collectors.groupingBy(book -> (book.getYear() / 10) * 10));
        
        System.out.println("\nLibros por década:");
        booksByDecade.forEach((decade, books) -> {
            System.out.printf("  %ds: %d libros%n", decade, books.size());
        });
    }
    
    /**
     * Prompt: Método que demuestra operaciones con archivos
     * Debe incluir escritura, lectura y procesamiento de JSON
     */
    private static void demonstrateFileOperations() throws IOException {
        System.out.println("\n💾 Ejemplo 3: Operaciones con archivos");
        
        Path tempDir = Paths.get(System.getProperty("java.io.tmpdir"));
        Path dataFile = tempDir.resolve("copilot_example.txt");
        Path jsonFile = tempDir.resolve("copilot_data.json");
        
        // Escribir datos a archivo
        List<String> lines = Arrays.asList(
            "Línea 1: Ejemplo con GitHub Copilot",
            "Línea 2: Procesamiento de archivos en Java",
            "Línea 3: Streams y NIO.2",
            "Línea 4: Programación funcional"
        );
        
        Files.write(dataFile, lines);
        System.out.println("Archivo escrito: " + dataFile);
        
        // Leer y procesar archivo
        Map<String, Object> fileStats = Files.lines(dataFile)
            .collect(HashMap::new,
                (map, line) -> {
                    map.put("totalLines", (Integer) map.getOrDefault("totalLines", 0) + 1);
                    map.put("totalChars", (Integer) map.getOrDefault("totalChars", 0) + line.length());
                    map.put("longestLine", Math.max(
                        (Integer) map.getOrDefault("longestLine", 0), line.length()));
                },
                (map1, map2) -> {}
            );
        
        System.out.println("Estadísticas del archivo: " + fileStats);
        
        // Crear JSON simple (sin librerías externas)
        String jsonData = String.format(
            "{\"timestamp\":\"%s\",\"stats\":%s,\"success\":true}",
            LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            fileStats.toString().replace("=", ":")
        );
        
        Files.write(jsonFile, jsonData.getBytes());
        System.out.println("JSON escrito: " + jsonFile);
        
        // Limpiar archivos temporales
        Files.deleteIfExists(dataFile);
        Files.deleteIfExists(jsonFile);
    }
    
    /**
     * Prompt: Método que demuestra uso del HttpClient moderno de Java 11+
     * Debe hacer peticiones GET y POST con manejo de errores
     */
    private static void demonstrateHttpClient() throws Exception {
        System.out.println("\n🌐 Ejemplo 4: HTTP Client");
        
        HttpClient client = HttpClient.newHttpClient();
        
        // GET request
        HttpRequest getRequest = HttpRequest.newBuilder()
            .uri(URI.create("https://httpbin.org/json"))
            .header("Accept", "application/json")
            .GET()
            .build();
        
        CompletableFuture<HttpResponse<String>> getResponse = client.sendAsync(
            getRequest, HttpResponse.BodyHandlers.ofString());
        
        getResponse.thenAccept(response -> {
            System.out.println("GET Response Status: " + response.statusCode());
            System.out.println("Response Body (preview): " + 
                response.body().substring(0, Math.min(100, response.body().length())) + "...");
        }).join();
        
        // POST request
        String postData = "{\"message\":\"Hello from Java Copilot!\",\"timestamp\":\"" + 
            LocalDateTime.now() + "\"}";
        
        HttpRequest postRequest = HttpRequest.newBuilder()
            .uri(URI.create("https://httpbin.org/post"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(postData))
            .build();
        
        CompletableFuture<HttpResponse<String>> postResponse = client.sendAsync(
            postRequest, HttpResponse.BodyHandlers.ofString());
        
        postResponse.thenAccept(response -> {
            System.out.println("POST Response Status: " + response.statusCode());
            System.out.println("Request echoed successfully");
        }).join();
    }
    
    /**
     * Prompt: Método que demuestra programación funcional en Java
     * Debe incluir Function, Supplier, Consumer y method references
     */
    private static void demonstrateFunctionalProgramming() {
        System.out.println("\n⚡ Ejemplo 5: Programación funcional");
        
        // Functions
        Function<String, String> upperCaseFunction = String::toUpperCase;
        Function<String, Integer> lengthFunction = String::length;
        Function<String, Integer> upperAndLength = upperCaseFunction.andThen(lengthFunction);
        
        List<String> words = Arrays.asList("java", "copilot", "programming", "functional");
        
        // Usar functions con streams
        Map<String, Integer> wordLengths = words.stream()
            .collect(Collectors.toMap(
                upperCaseFunction,
                lengthFunction
            ));
        
        System.out.println("Palabras y sus longitudes:");
        wordLengths.forEach((word, length) -> 
            System.out.printf("  %s: %d caracteres%n", word, length));
        
        // Memoization example
        Function<Integer, Integer> expensiveCalculation = CopilotExamples::fibonacci;
        Function<Integer, Integer> memoizedFib = memoize(expensiveCalculation);
        
        System.out.println("\nSecuencia Fibonacci (con memoization):");
        IntStream.rangeClosed(1, 10)
            .forEach(i -> System.out.printf("F(%d) = %d ", i, memoizedFib.apply(i)));
        System.out.println();
        
        // Validation chain
        List<Function<String, Optional<String>>> validators = Arrays.asList(
            CopilotExamples::validateNotEmpty,
            CopilotExamples::validateEmail,
            CopilotExamples::validateEmailDomain
        );
        
        String testEmail = "user@example.com";
        Optional<String> validationResult = validators.stream()
            .map(validator -> validator.apply(testEmail))
            .filter(Optional::isPresent)
            .findFirst()
            .orElse(Optional.empty());
        
        System.out.println("\nValidación de email '" + testEmail + "': " + 
            (validationResult.isPresent() ? validationResult.get() : "✅ Válido"));
    }
    
    /**
     * Prompt: Método que implementa fibonacci con recursión simple
     * Para demostrar memoization
     */
    private static Integer fibonacci(Integer n) {
        if (n <= 1) return n;
        return fibonacci(n - 1) + fibonacci(n - 2);
    }
    
    /**
     * Prompt: Función que implementa memoization para cualquier función
     * Usa un cache concurrente para almacenar resultados
     */
    private static <T, R> Function<T, R> memoize(Function<T, R> function) {
        Map<T, R> cache = new HashMap<>();
        return input -> {
            return cache.computeIfAbsent(input, function);
        };
    }
    
    // Validators para demostrar programación funcional
    private static Optional<String> validateNotEmpty(String value) {
        return value == null || value.trim().isEmpty() ? 
            Optional.of("El valor no puede estar vacío") : Optional.empty();
    }
    
    private static Optional<String> validateEmail(String email) {
        Pattern emailPattern = Pattern.compile("^[A-Za-z0-9+_.-]+@(.+)$");
        return emailPattern.matcher(email).matches() ? 
            Optional.empty() : Optional.of("Formato de email inválido");
    }
    
    private static Optional<String> validateEmailDomain(String email) {
        return email.endsWith("@spam.com") ? 
            Optional.of("Dominio no permitido") : Optional.empty();
    }
}

/**
 * Clase Person generada con GitHub Copilot
 * Prompt: Clase inmutable que representa una persona empleada
 * Debe incluir validation y métodos utilitarios
 */
class Person {
    private final String name;
    private final int age;
    private final String department;
    private final double salary;
    
    public Person(String name, int age, String department, double salary) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("El nombre no puede estar vacío");
        }
        if (age < 18 || age > 100) {
            throw new IllegalArgumentException("La edad debe estar entre 18 y 100");
        }
        if (salary < 0) {
            throw new IllegalArgumentException("El salario no puede ser negativo");
        }
        
        this.name = name.trim();
        this.age = age;
        this.department = department;
        this.salary = salary;
    }
    
    // Getters
    public String getName() { return name; }
    public int getAge() { return age; }
    public String getDepartment() { return department; }
    public double getSalary() { return salary; }
    
    @Override
    public String toString() {
        return String.format("Person{name='%s', age=%d, department='%s', salary=%.2f}", 
            name, age, department, salary);
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Person person = (Person) obj;
        return age == person.age && 
               Double.compare(person.salary, salary) == 0 && 
               Objects.equals(name, person.name) && 
               Objects.equals(department, person.department);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(name, age, department, salary);
    }
}

/**
 * Clase Book generada con GitHub Copilot
 * Prompt: Clase que representa un libro con ISBN, título, autor y año
 * Debe incluir validación de ISBN y métodos de búsqueda
 */
class Book {
    private final String title;
    private final String author;
    private final String isbn;
    private final int year;
    
    public Book(String title, String author, String isbn, int year) {
        this.title = Objects.requireNonNull(title, "El título no puede ser null");
        this.author = Objects.requireNonNull(author, "El autor no puede ser null");
        this.isbn = validateISBN(isbn);
        this.year = year;
    }
    
    private String validateISBN(String isbn) {
        if (isbn == null || isbn.trim().isEmpty()) {
            throw new IllegalArgumentException("ISBN no puede estar vacío");
        }
        // Simplificada: validar formato básico
        String cleanISBN = isbn.replaceAll("[^0-9X]", "");
        if (cleanISBN.length() != 10 && cleanISBN.length() != 13) {
            throw new IllegalArgumentException("ISBN debe tener 10 o 13 dígitos");
        }
        return isbn;
    }
    
    // Getters
    public String getTitle() { return title; }
    public String getAuthor() { return author; }
    public String getIsbn() { return isbn; }
    public int getYear() { return year; }
    
    public boolean matchesKeyword(String keyword) {
        String lowerKeyword = keyword.toLowerCase();
        return title.toLowerCase().contains(lowerKeyword) || 
               author.toLowerCase().contains(lowerKeyword);
    }
    
    @Override
    public String toString() {
        return String.format("Book{title='%s', author='%s', year=%d}", title, author, year);
    }
}

/**
 * Clase Library generada con GitHub Copilot
 * Prompt: Clase que maneja una colección de libros
 * Debe incluir métodos de búsqueda, filtrado y estadísticas
 */
class Library {
    private final String name;
    private final List<Book> books;
    
    public Library(String name) {
        this.name = Objects.requireNonNull(name, "El nombre no puede ser null");
        this.books = new ArrayList<>();
    }
    
    public void addBook(Book book) {
        Objects.requireNonNull(book, "El libro no puede ser null");
        books.add(book);
    }
    
    public List<Book> getBooks() {
        return new ArrayList<>(books); // Defensive copy
    }
    
    public String getName() { return name; }
    
    public int getTotalBooks() { return books.size(); }
    
    public List<Book> searchByKeyword(String keyword) {
        return books.stream()
            .filter(book -> book.matchesKeyword(keyword))
            .collect(Collectors.toList());
    }
    
    public List<Book> getBooksByAuthor(String author) {
        return books.stream()
            .filter(book -> book.getAuthor().equalsIgnoreCase(author))
            .collect(Collectors.toList());
    }
    
    public OptionalDouble getAverageYear() {
        return books.stream()
            .mapToInt(Book::getYear)
            .average();
    }
}